riptide-opentelemetry/src/main/java/org/zalando/riptide/opentelemetry/OpenTelemetryPlugin.java

```diff
@@ -0,0 +1,82 @@
+package org.zalando.riptide.opentelemetry;
+
+import com.google.common.collect.Multimaps;
+import io.opentelemetry.api.GlobalOpenTelemetry;
+import io.opentelemetry.api.trace.Span;
+import io.opentelemetry.api.trace.SpanKind;
+import io.opentelemetry.api.trace.Tracer;
+import io.opentelemetry.context.Context;
+import io.opentelemetry.context.Scope;
+import io.opentelemetry.context.propagation.TextMapPropagator;
+import org.springframework.http.client.ClientHttpResponse;
+import org.zalando.riptide.Attribute;
+import org.zalando.riptide.Plugin;
+import org.zalando.riptide.RequestArguments;
+import org.zalando.riptide.RequestExecution;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+public class OpenTelemetryPlugin implements Plugin {
+
+    public static final Attribute<String> OPERATION_NAME = Attribute.generate();
+
+    private final Tracer tracer;
+    private final TextMapPropagator propagator;
+
+    private final Attribute<Span> internalSpan = Attribute.generate();
+
+    public OpenTelemetryPlugin(final Tracer tracer) {
+        this.tracer = tracer;
+        this.propagator = GlobalOpenTelemetry.getPropagators().getTextMapPropagator();
+    }
+
+    @Override
+    public RequestExecution aroundAsync(@Nonnull final RequestExecution execution) {
```

Comment:
You're correct. The method itself is called from the calling thread. The returned lambda is called on the IO thread though.

The OpenTracing plugin should give you a good blueprint for how to transfer the span over.

Reply:
You're correct. 
a) Method called from initial thread, so it's proper place to grab parent context. 
Still creation of spans in the method isn't right place, since backup / retry requests wouldn't be covered. 
b) not sure but I want to look on tagging retries/backup requests etc

Reply:
Ok, so if I understand this correctly (and I acknowledge I may be missing something), we are not passing the span to the thread executing the request, rather opting to work with the span and the request's response that is returned from that thread. Did I get it right?

Reply:
Almost. We're not passing span from caller method, but context. Spans will be created in the threads that will execute calls

Reply:
Looking at the `trace` method the context is used to create the span wrapping the HTTP call (line 68). That span then becomes the current span (line 76), and that context is injected into the headers (line 77), which are then added to `arguments` (line 80).
This means that the executing thread has access to the context through those headers, and if needed will have to build the context with them. I guess I was expecting something like we have in the [OpenTracing plugin](https://github.com/zalando/riptide/blob/main/riptide-opentracing/src/main/java/org/zalando/riptide/opentracing/OpenTracingPlugin.java#L174), that passes the span itself (although, with OpenTelemetry it would make sense to just pass the context. WDYT?

Reply:
OpenTracing plugin just adds layer of abstraction by providing two kinds of injectors, headers and noop. But effectively injection to headers happens during execution of trace method via `injection.inject(tracer, arguments, context)`. Current implementation doesn't have support for such injectors, and I'm not sure if needed.